php_phongo.c

```diff
@@ -1879,17 +1333,94 @@ static bool php_phongo_apply_wc_options_to_uri(mongoc_uri_t *uri, bson_t *option
 	return true;
 } /* }}} */
 
+static bool php_phongo_apply_ssl_opts(mongoc_client_t *client, zval *driverOptions TSRMLS_DC)
+{
+	mongoc_ssl_opt_t ssl_opt = {0};
+	zend_bool free_pem_file = 0, free_pem_pwd = 0, free_ca_file = 0, free_ca_dir = 0;
+	zval *sslOptions = driverOptions;
+	int plen;
+
+	if (!sslOptions) {
+		return true;
+	}
+
+	if (php_array_existsc(sslOptions, "context")) {
+		php_stream_context *context;
+		zval *contextOption = php_array_fetchc(driverOptions, "context");
+
+		/* TODO: Do we want to check global stream options, too? */
+		context = php_stream_context_from_zval(contextOption, 1);
+
+		if (!context) {
+			phongo_throw_exception(PHONGO_ERROR_INVALID_ARGUMENT TSRMLS_CC, "\"context\" driver option is not a valid Stream-Context resource");
+			return false;
+		}
+
+#if PHP_VERSION_ID >= 70000
+		sslOptions = php_array_fetchc_array(&context->options, "ssl");
+#else
+		sslOptions = php_array_fetchc_array(context->options, "ssl");
+#endif
+
+		if (!sslOptions) {
+			phongo_throw_exception(PHONGO_ERROR_INVALID_ARGUMENT TSRMLS_CC, "Stream-Context resource does not contain \"ssl\" options array");
+			return false;
+		}
+	}
+
+	if (php_array_existsc(sslOptions, "verify_hostname")) {
+		ssl_opt.allow_invalid_hostname = !php_array_fetchc_bool(sslOptions, "verify_hostname");
+	}
+
+	if (php_array_existsc(sslOptions, "allow_self_signed")) {
+		ssl_opt.weak_cert_validation = php_array_fetchc_bool(sslOptions, "allow_self_signed");
+	}
+
+	if (php_array_existsc(sslOptions, "local_cert")) {
+		ssl_opt.pem_file = php_array_fetchc_string(sslOptions, "local_cert", &plen, &free_pem_file);
+	}
+
+	if (php_array_existsc(sslOptions, "passphrase")) {
+		ssl_opt.pem_pwd = php_array_fetchc_string(sslOptions, "passphrase", &plen, &free_pem_pwd);
+	}
+
+	if (php_array_existsc(sslOptions, "cafile")) {
+		ssl_opt.ca_file = php_array_fetchc_string(sslOptions, "cafile", &plen, &free_ca_file);
+	}
+
+	if (php_array_existsc(sslOptions, "capath")) {
+		ssl_opt.ca_dir = php_array_fetchc_string(sslOptions, "capath", &plen, &free_ca_dir);
```

Comment:
Am I correct in assuming that only different here is customizing the entire CA path? All three will allow a custom CA file, which I suppose most users can stick to if they have a special cert.

Would the default CA path for each generally be sufficient? Since we won't be using PHP's OpenSSL extension at all, I'm on the fence about whether we should even default to `openssl.capath`.